src/shogun/lib/any.h

```diff
@@ -439,102 +433,80 @@ namespace shogun {
-		auto compare_impl(general, const T& lhs, const T& rhs)
-		    -> decltype(lhs == rhs)
-		{
-			return compare_impl_eq(lhs, rhs);
-		}
-
-		template <class T>
-		auto compare_impl(more_important, const T& lhs, const T& rhs)
-		    -> decltype(lhs.equals(rhs))
-		{
-			return lhs.equals(rhs);
-		}
-
-		template <class T>
-		auto compare_impl(maybe_most_important, T* lhs, T* rhs)
-		    -> decltype(lhs->equals(rhs))
+		bool compare(const T& lhs, const T& rhs)
 		{
-			if (lhs && rhs)
-				return lhs->equals(rhs);
-			else if (!lhs && !rhs)
-				return true;
-			else
-				return false;
-		}
-
-		template <class T>
-		inline bool compare(const T& lhs, const T& rhs)
-		{
-			return compare_impl(maybe_most_important(), lhs, rhs);
-		}
-
-
-		template <class T1, class T2>
-		inline bool compare(const std::pair<T1, T2>& lhs, const std::pair<T1, T2>& rhs)
-		{
-			return (compare_impl(maybe_most_important(), lhs.first, rhs.first) &&
-				compare_impl(maybe_most_important(), lhs.second, rhs.second));
-		}
-
-		template <class T>
-		bool compare_impl(
-		    maybe_most_important, const std::function<T()>& lhs,
-		    const std::function<T()>& rhs)
-		{
-			return compare(lhs(), rhs());
-		}
-
-		template <class T,
-			std::enable_if_t<utils::is_container<T>::value>* = nullptr>
-		bool compare_impl(
-		    maybe_most_important, const T& lhs, const T& rhs)
-		{
-			if (lhs.size() != rhs.size())
+			if constexpr (std::is_same<T, float32_t>::value)
+			{
+				return compare_impl_eq(lhs, rhs);
+			}
+			if constexpr (std::is_same<T, float64_t>::value)
+			{
+				return compare_impl_eq(lhs, rhs);
+			}
+			if constexpr (std::is_same<T, floatmax_t>::value)
+			{
+				return compare_impl_eq(lhs, rhs);
+			}
+			if constexpr (std::is_same<T, complex128_t>::value)
 			{
-				return false;
+				return compare_impl_eq(lhs, rhs);
 			}
-			for (auto l = lhs.cbegin(), r = rhs.cbegin(); l != lhs.cend();
-			     ++l, ++r)
+			if constexpr (traits::is_container<T>::value)
 			{
-				if (!compare(*l, *r))
+				if (lhs.size() != rhs.size())
 				{
 					return false;
 				}
+				for (auto l = lhs.cbegin(), r = rhs.cbegin(); l != lhs.cend(); ++l, ++r)
+				{
+					if (!compare(*l, *r))
+					{
+						return false;
+					}
+				}
 			}
-
-			return true;
-		}
-
-		template <class T>
-		inline size_t hash_impl(general, const T&)
-		{
-			return 0;
-		}
-
-		template <class T>
-		inline auto hash_impl(more_important, const T& value)
-		    -> decltype(std::hash<T>{}(value))
-		{
-			return std::hash<T>{}(value);
-		}
-
-		template <class T,
-			std::enable_if_t<utils::is_container<T>::value>* = nullptr>
-		size_t hash(maybe_most_important, const T& value)
-		{
-			size_t result = 0;
-			for (const auto& it: value) {
-				result ^= hash(it);
+			if constexpr (traits::is_pair<T>::value)
+			{
+				return compare(lhs.first, rhs.first) && compare(lhs.second, rhs.second);
+			}
+			if constexpr (traits::is_functional<T>::value)
+			{
+				return compare(lhs(), rhs());
```

Comment:
Do you have an example?

I see no problem having `void` so far.

Reply:
what works now and what i wish could work:
```cpp
class A
{
public:
    A()
    {
        add_func(&A::do_something1); // works
        add_func(&A::do_something2); // doesn't work
    }

    template <typename T, typename S>
    void add_func(T (S::*method)())
    {
        std::function<T()> bind_method =
            std::bind(method, dynamic_cast<S*>(this));
        m_func = make_any(bind_method);
    }

    int do_something1()
    {
       m_a = 1;
        return m_a;
    }
    void do_something2()
    {
        m_a = 1;
    }
    int run()
    {
        return any_cast<int>(m_func); // works
        // any_cast<void>(m_func); // doesn't work
        // return m_a; // if second one would work we would return m_a
    }
private:
    int m_a;
    Any m_func;
};
```
make_any with do_something2 causes issues because in this line you're comparing void vs void. And then I get an error like this:
```
In file included from /home/gf712/shogun/src/shogun/base/AnyParameter.h:11:0,
                 from /home/gf712/shogun/src/shogun/base/SGObject.h:14,
                 from /home/gf712/shogun/src/shogun/io/File.h:14,
                 from /home/gf712/shogun/src/shogun/features/Features.h:15,
                 from iterator_test.cpp:1:
/home/gf712/shogun/src/shogun/lib/any.h: In instantiation of ‘bool shogun::any_detail::compare_impl(shogun::any_detail::maybe_most_important, const std::function<T()>&, const std::function<T()>&) [with T = void]’:
/home/gf712/shogun/src/shogun/lib/any.h:470:23:   required from ‘bool shogun::any_detail::compare(const T&, const T&) [with T = std::function<void()>]’
/home/gf712/shogun/src/shogun/lib/any.h:930:18:   required from ‘bool shogun::PointerValueAnyPolicy<T>::equals(const void*, const void*) const [with T = std::function<void()>]’
iterator_test.cpp:116:1:   required from here
/home/gf712/shogun/src/shogun/lib/any.h:486:18: error: invalid use of void expression
    return compare(lhs(), rhs());
           ~~~~~~~^~~~~~~~~~~~~~
```

Reply:
Oh ok. Then I need to change the traits things into checking if it is not void.

Reply:
Cool then we could remove the Boolean hack in “run”

Reply:
Yes exactly what I was thinking!

Reply:
Added a check and a test in 773bb59